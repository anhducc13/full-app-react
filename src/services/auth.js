import axios from 'axios';
const URL = process.env.REACT_APP_API_URL

const register = params => {
    return axios.post(`${URL}/auth/register`, params)
}

const verify = (accessToken) => {
    return axios.get(`${URL}/auth/verify/${accessToken}`)
}

const login = (params) => {
    return axios.post(`${URL}/auth/login`, params)
}

const logout = (accessToken) => {
    const reqConfig = {
        "headers": {
            "Authorization": `Bearer ${accessToken}`,
        }

    }
    return axios.get(`${URL}/auth/logout`, reqConfig)
}

const updatePassword = (params, accessToken) => {
    const reqConfig = {
        "headers": {
            "Authorization": `Bearer ${accessToken}`,
        }
    }
    return axios.post(`${URL}/auth/updatePassword`, params, reqConfig)
}

const forgotPassword = params => {
    return axios.post(`${URL}/auth/forgotPassword`, params)
}

export default {
    login,
    logout,
    register,
    verify,
    updatePassword,
    forgotPassword
}