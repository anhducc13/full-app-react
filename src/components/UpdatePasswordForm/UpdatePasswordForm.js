import React, { useState } from 'react';
import { setGlobal } from 'reactn';
import Avatar from '@material-ui/core/Avatar';
import CssBaseline from '@material-ui/core/CssBaseline';
import Link from '@material-ui/core/Link';
import Grid from '@material-ui/core/Grid';
import LockOutlinedIcon from '@material-ui/icons/LockOutlined';
import Typography from '@material-ui/core/Typography';
import { makeStyles } from '@material-ui/core/styles';
import Container from '@material-ui/core/Container';
import { Link as RouteLink } from 'react-router-dom';
import { Button as ButtonCustom } from '../shared/Button';
import { InputText } from '../shared/InputText';
import Swal from 'sweetalert2';
import { authService } from 'services';
import { validatePassword } from 'helpers/validators';

const useStyles = makeStyles(theme => ({
  '@global': {
    body: {
      backgroundColor: theme.palette.common.white,
    },
  },
  paper: {
    marginTop: theme.spacing(2),
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
  },
  avatar: {
    margin: theme.spacing(1),
    backgroundColor: theme.palette.secondary.main,
    width: 60,
    height: 60,
  },
  form: {
    width: '100%', // Fix IE 11 issue.
    marginTop: theme.spacing(3),
  },
}));

export const UpdatePasswordForm = (props) => {
  const classes = useStyles();

  const initTextField = {
    value: '',
    error: '',
    valid: false,
  };
  const [passwordField, setPasswordField] = useState(initTextField);
  const [newPasswordField, setNewPasswordField] = useState(initTextField);
  const [confirmNewPasswordField, setConfirmNewPasswordField] = useState(initTextField);

  const onChangePassword = (e) => {
    let {value, error, valid} = passwordField;
    value = e.target.value;
    const outputValidate = validatePassword({value, error, valid});
    setPasswordField(outputValidate);
  }

  const onChangeNewPassword = (e) => {
    let {value, error, valid} = newPasswordField;
    value = e.target.value;
    const outputValidate = validatePassword({value, error, valid});
    setNewPasswordField(outputValidate);
  }

  const onChangeConfirmNewPassword = (e) => {
    let {value, error, valid} = confirmNewPasswordField;
    value = e.target.value;

    if (value !== newPasswordField.value) {
      error = 'Mật khẩu không khớp!';
      valid = false;
    } else {
      error = '';
      valid = true;
    }
    setConfirmNewPasswordField({value, error, valid});
  }

  const onSubmitForm = (e) => {
    e.preventDefault()
    const payload = {
      "old_password": passwordField.value,
      "new_password": newPasswordField.value,
    }
    setGlobal({
      loading: true,
    })
    const accessToken = JSON.parse(localStorage.getItem('userInfoLogin'))['access_token'];
    authService.updatePassword(payload, accessToken)
      .then(() => {
        setGlobal({
          loading: false,
        })
        Swal.fire(
          'Thành công!',
          'Mật khẩu của bạn đã được thay đổi',
          'success'
        )
          .then(() => {
            props.history.push('/home')
          })
      })
      .catch(err => {
        setGlobal({
          loading: false,
        })
        if (err.response && err.response.status === 401) {
          Swal.fire(
            'Có lỗi xảy ra!',
            'Từ chối truy cập',
            'error'
          )
            .then(() => {
              props.history.push('/error')
            })
        } else if (err.response && err.response.data.message === '400 Bad Request: password is not match') {
          Swal.fire(
            'Có lỗi xảy ra!',
            'Mật khẩu nhập không chính xác',
            'error'
          )
        } else if (err.response && err.response.data.message === '400 Bad Request: two password is duplicate') {
          Swal.fire(
            'Có lỗi xảy ra!',
            'Mật khẩu mới không được trùng với mật khẩu cũ',
            'error'
          )
        } else {
          props.history.push('/error')
        }
      })
  }

  return (
    <Container component="main" maxWidth="sm">
      <CssBaseline />
      <div className={classes.paper}>
        <Avatar className={classes.avatar}>
          <LockOutlinedIcon />
        </Avatar>
        <Typography component="h1" variant="h4">
          Reset Password
        </Typography>
        <form className={classes.form} noValidate onSubmit={onSubmitForm}>
          <InputText
            textField={passwordField}
            name="password"
            label="Password"
            type="password"
            onChange={onChangePassword}
          />
          <InputText
            textField={newPasswordField}
            name="new-password"
            label="New Password"
            type="password"
            onChange={onChangeNewPassword}
          />
          <InputText
            textField={confirmNewPasswordField}
            name="confirm-new-password"
            label="Confirm New Password"
            type="password"
            onChange={onChangeConfirmNewPassword}
          />
          <ButtonCustom
            displayText="Update"
            disabled={
              !passwordField.valid || !newPasswordField.valid || 
              !confirmNewPasswordField.valid
            }
          />
          <Grid container>
            <Link to="/home" variant="body2" component={RouteLink}>
              Back to home
            </Link>
          </Grid>
        </form>
      </div>
    </Container>
  );
}