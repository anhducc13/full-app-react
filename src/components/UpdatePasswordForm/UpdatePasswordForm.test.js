import React from 'react';
import { mount, shallow } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import { UpdatePasswordForm } from './UpdatePasswordForm';
import { authService } from '../../services';

describe('Update Password Form Render', () => {
  it('should render correctly', () => {
    const component = shallow(<UpdatePasswordForm />);
    expect(component).toMatchSnapshot();
  });
});

describe('Update Password Form validate password', () => {
  it('validate password test case 1', () => {
    const component = mount(
      <MemoryRouter>
        <UpdatePasswordForm />
      </MemoryRouter>
    );
    const targetInput = {
      target: {
        value: ""
      }
    }
    component.find('input[name="password"]').simulate('change', targetInput);
    expect(component.find('p#password-helper-text').length)
      .toBe(1);
    component.unmount()
  })

  it('validate password test case 2', () => {
    const component = mount(
      <MemoryRouter>
        <UpdatePasswordForm />
      </MemoryRouter>
    );
    const targetInput = {
      target: {
        value: " "
      }
    }
    component.find('input[name="password"]').simulate('change', targetInput);
    expect(component.find('p#password-helper-text').text())
      .toBe('Mật khẩu không được chứa khoảng trắng');
    component.unmount()
  })

  it('validate confirm new password', () => {
    const component = mount(
      <MemoryRouter>
        <UpdatePasswordForm />
      </MemoryRouter>
    );
    const targetInput = {
      target: {
        value: "Anhducc13"
      }
    }
    component.find('input[name="new-password"]').simulate('change', targetInput);
    component.find('input[name="confirm-new-password"]').simulate('change', `${targetInput} `) ;
    expect(component.find('p#confirm-new-password-helper-text').text())
        .toBe('Mật khẩu không khớp')
    component.unmount()
  })
});

describe('Update password form submit', () => {
  it('Button submit disable with invalid data', () => {
    const component = mount(
      <MemoryRouter>
        <UpdatePasswordForm />
      </MemoryRouter>
    );
    component.find('input[name="password"]').simulate('change', { target: { value: '' } });
    expect(component.find('button[type="submit"]').prop('disabled')).toBeTruthy();
    component.unmount();
  })

  it('Button submit enable with expect data', () => {
    const component = mount(
      <MemoryRouter>
        <UpdatePasswordForm />
      </MemoryRouter>
    );
    component.find('input[name="password"]').simulate('change', { target: { value: 'Anhducc13' } });
    component.find('input[name="new-password"]').simulate('change', { target: { value: 'Anhducc14' } });
    component.find('input[name="confirm-new-password"]').simulate('change', { target: { value: 'Anhducc14' } });
    expect(component.find('button[type="submit"]').prop('disabled')).toBeFalsy()
    component.unmount();
  })

  it('Submit form, call api with valid input', () => {
    const component = mount(
      <MemoryRouter>
        <UpdatePasswordForm />
      </MemoryRouter>
    );
    const value = {
      'old_password': 'Anhducc13',
      'new_password': 'Anhducc14'
    }
    const userInfoLogin = {
        username: 'anhducc13',
        timeExpired: Date.now(),
        accessToken: 'asfdfbrfdfgsdcsdgsdfgsg'
    }
    localStorage.setItem('userInfoLogin', JSON.stringify(userInfoLogin))
    component.find('input[name="password"]').simulate('change', { target: { value: value['old_password'] } });
    component.find('input[name="new-password"]').simulate('change', { target: { value: value['new_password'] } });
    component.find('input[name="confirm-new-password"]').simulate('change', { target: { value: value['new_password'] } });
    const updatePasswordFn = jest.spyOn(authService, 'updatePassword').mockImplementation(() => Promise.resolve({ status: 200 }));
    component.find('form').simulate('submit');
    expect(updatePasswordFn).toHaveBeenCalled();
    expect(updatePasswordFn).toHaveBeenCalledWith(value, userInfoLogin.accessToken);
    component.unmount();
  })
});
