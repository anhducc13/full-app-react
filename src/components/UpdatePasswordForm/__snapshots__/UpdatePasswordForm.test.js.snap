// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Update Password Form Render should render correctly 1`] = `ShallowWrapper {}`;
